constraints/miden-vm/stack.air

```diff
@@ -0,0 +1,715 @@
+    enf f_movup6(b) * (s[i + 1]' - s[i]) = 0 for i in 0..6
+    enf f_movup6(b) * delta(s[i]) = 0 for i in 7..16
+
+ev movdn6(main: [s[16]], aux: [b[7]]):
+    enf f_movdn6(b) * (s[6]' - s[0]) = 0
+    enf f_movdn6(b) * (s[i]' - s[i + 1]) = 0 for i in 0..6
+    enf f_movdn6(b) * delta(s[i]) = 0 for i in 7..16
+
+ev movup7(main: [s[16]], aux: [b[7]]):
+    enf f_movup7(b) * (s[0]' - s[7]) = 0
+    enf f_movup7(b) * (s[i + 1]' - s[i]) = 0 for i in 0..7
+    enf f_movup7(b) * delta(s[i]) = 0 for i in 8..16
+
+ev movdn7(main: [s[16]], aux: [b[7]]):
+    enf f_movdn7(b) * (s[7]' - s[0]) = 0
+    enf f_movdn7(b) * (s[i]' - s[i + 1]) = 0 for i in 0..7
+    enf f_movdn7(b) * delta(s[i]) = 0 for i in 8..16
+
+ev swapw(main: [s[16]], aux: [b[7]]):
+    enf f_swapw(b) * (s[i]' - s[i + 4]) = 0 for i in 0..4
+    enf f_swapw(b) * (s[i + 4]' - s[i]) = 0 for i in 0..4
+    enf f_swapw(b) * delta(s[i]) = 0 for i in 8..16
+
+ev ext2mul(main: [s[16]], aux: [b[7]]):
+    enf f_ext2mul(b) * (s[0]' - s[0]) = 0
+    enf f_ext2mul(b) * (s[1]' - s[1]) = 0
+    enf f_ext2mul(b) * (s[2]' - (s[0] + s[1]) * (s[2] + s[3]) + s[0] * s[2]) = 0
+    enf f_ext2mul(b) * (s[3]' - s[1] * s[3] + 2 * s[0] * s[2]) = 0
+    enf f_ext2mul(b) * delta(s[i]) = 0 for i in 4..16
+
+ev movup8(main: [s[16]], aux: [b[7]]):
+    enf f_movup8(b) * (s[0]' - s[8]) = 0
+    enf f_movup8(b) * (s[i + 1]' - s[i]) = 0 for i in 0..8
+    enf f_movup8(b) * delta(s[i]) = 0 for i in 9..16
+
+ev movdn8(main: [s[16]], aux: [b[7]]):
+    enf f_movdn8(b) * (s[8]' - s[0]) = 0
+    enf f_movdn8(b) * (s[i]' - s[i + 1]) = 0 for i in 0..8
+    enf f_movdn8(b) * delta(s[i]) = 0 for i in 9..16
+
+ev swapw2(main: [s[16]], aux: [b[7]]):
+    enf f_swapw2(b) * (s[i]' - s[i + 8]) = 0 for i in 0..4
+    enf f_swapw2(b) * (s[i + 8]' - s[i]) = 0 for i in 0..4
+    enf f_swapw2(b) * delta(s[i]) = 0 for i in 4..8
+    enf f_swapw2(b) * delta(s[i]) = 0 for i in 12..16
+
+# Should we enforce that elements in position grater than 15 do not change? 
+ev swapw3(main: [s[16]], aux: [b[7]]):
+    enf f_swapw3(b) * (s[i]' - s[i + 12]) = 0 for i in 0..4
+    enf f_swapw3(b) * (s[i + 12]' - s[i]) = 0 for i in 0..4
+    enf f_swapw3(b) * delta(s[i]) = 0 for i in 4..12
+
+# Should we enforce that elements in position grater than 15 do not change? 
+ev swapdw(main: [s[16]], aux: [b[7]]):
+    enf f_swapdw(b) * (s[i]' - s[i + 8]) = 0 for i in 0..8
+    enf f_swapdw(b) * (s[i + 8]' - s[i]) = 0 for i in 0..8
+
+
+# Left stack shift operationds
+
+ev assert(main: [s[16]], aux: [b[7]]):
+    enf f_assert(b) * (s[0] - 1) = 0
+    enf f_assert(b) * (s[i]' - s[i + 1]) = 0 for i in 0..15
+
+ev eq(main: [s[16]], aux: [b[7], helper]):
+    enf f_eq(b) * (s[0]' * (s[0] - s[1])) = 0
+    enf f_eq(b) * (s[0]' - (1 - (s[0] - s[1]) * helper)) = 0
+    enf f_eq(b) * (s[i]' - s[i + 1]) = 0 for i in 1..15
+
+ev add(main: [s[16]], aux: [b[7]]):
+    enf f_add(b) * (s[0]' - (s[0] + s[1])) = 0
+    enf f_eq(b) * (s[i]' - s[i + 1]) = 0 for i in 1..15
```

Comment:
I would not make evaluators aware of the flag values - this could be handled by the calling evaluator as a part of `match enf` statement. So, I would rewrite the evaluators in this file to work like that. For example, the evaluator for `add` operation could look like this:
```
ev add(main: [s[16]]):
  enf s[0]' = s[0] + s[1]
  enf s[i]' = s[i + 1] for i in 1..15
```
Also, do we already support computed indexes (i.e., `s[i + 1]`)? Or is this something planned?

Reply:
And to provide more details about why I meant by "calling evaluator": I'm imagining that we'll have a vector of 88 mutually-exclusive flags and then we'll select based on that. If we had only 4 operations, this could look something like this:
```
# this returns a vector of mutually exclusive operation flags
let op_flags = compute_op_flags()
match enf
  op_add(stack_top) when op_flags[0]
  op_neg(stack_top) when op_flags[1]
  op_mul(stack_top) when op_flags[2]
  op_inv(stack_top) when op_flags[3]
```
In the above `stack_top` is a reference to the 16 columns representing 16 columns of the stack top.